### PR TITLE
feat: check user exists in AD then login or register him

### LINKED ISSUE #102: [FEATURE]: Azure authentication (sso)
## Summary

As an authenticated user I want to sign in with azure authentication

### PR DESCRIPTION
**Pull request by nunocaseiro**
Relates to #102

## Proposed Changes

  - Endpoint and respective method to check if the user exists in the active directory
  - Endpoint and respective method to check if the user can login or must first be registered in our database
  - Cronjob to get a azure access token to be able to make requests to microsoft graph services

This pull request is intended to close #102

### CODE REVIEW

File: backend/src/modules/auth/shared/login.auth.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  if (jwt) {
+    return {
+      ...jwt,
+      email: user.email,
+      name: user.name,
+      strategy,
+      id: user._id,
+    };
+  }
+  return null;
```
Comment (rpvsilva) [inline backend/src/modules/auth/shared/login.auth.ts:22]:
```suggestion
  if (!jwt) return null;
  
  return {
    ...jwt,
    email: user.email,
    name: user.name,
    strategy,
    id: user._id,
  };
```

File: backend/src/modules/azure/services/auth.azure.service.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      name: `${given_name} ${family_name}`,
+    });
+    if (createdUser) return signIn(createdUser, this.getTokenService, 'azure');
+
+    return null;
+  }
+
+  async checkUserExistsInActiveDirectory(email: string) {
+    const { data } = await axios.get(
+      `https://graph.microsoft.com/v1.0/users?$search="mail:${email}" OR "displayName:${email}" OR "userPrincipalName:${email}"&$orderbydisplayName&$count=true`,
```
Comment (rpvsilva) [inline backend/src/modules/azure/services/auth.azure.service.ts:58]: Put this link into a function that receives an email 😉

File: backend/src/modules/azure/services/auth.azure.service.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    const user = data.value.find(
+      (userFound: AzureUserFound) =>
+        userFound.mail?.toLowerCase() === email.toLowerCase() ||
+        userFound.displayName?.toLowerCase() === email.toLowerCase() ||
+        userFound.userPrincipalName?.toLowerCase() === email.toLowerCase(),
+    );
+    if (user) return true;
+
+    return false;
```
Comment (rpvsilva) [inline backend/src/modules/azure/services/auth.azure.service.ts:75]:
```suggestion
    return !isEmpty(user)
```

File: backend/src/modules/azure/services/cron.azure.service.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    const config = {
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+    };
+
+    const { data: tokenData } = await axios.post(
+      `https://login.microsoftonline.com/${this.configService.get(
+        AZURE_TENANT_ID,
+      )!}/oauth2/v2.0/token`,
```
Comment (rpvsilva) [inline backend/src/modules/azure/services/cron.azure.service.ts:48]: Put this in one function that receives the tenant id

**Review by rpvsilva — CHANGES_REQUESTED** [review]
Some small changes but you are rocking it 🚀

File: backend/src/modules/azure/services/cron.azure.service.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  @Cron('0 45 * * * *')
+  async handleCron() {
+    await this.getAzureAccessToken();
+  }
+
+  public getToken() {
+    return this.azureAccessToken;
+  }
+
+  getAzureAccessToken = async () => {
```
Comment (rpvsilva) [inline backend/src/modules/azure/services/cron.azure.service.ts:29]: Just to be more coherent 😉 

```suggestion
  async getAzureAccessToken() {
```

**Comment by nunocaseiro** [comment]
> Some small changes but you are rocking it 🚀

Thanks bro :)

**Review by rpvsilva — APPROVED** [review]

**Pull request merged**